File: Python/CreationalDesignPattern/ProxyPatternDemo.py
# class ProxyPatternDemo:
#
#     # Subject
#     class Database(ABC):
#         @abstractmethod
#         def fetch_data(self):
#             pass
#
#     # Real Subject
#     class RealDatabase(Database):
#         def fetch_data(self):
#             print("Fetching confidential employee salary data...")
#
#     # Proxy
#     class DatabaseProxy(Database):
#         def __init__(self, role):
#             self.role = role
#             self.database = None
#
#         def fetch_data(self):
#             if self.role.upper() != "ADMIN":
#                 print("Access Denied! Only ADMIN can access the database.")
#                 return
#
#             # Lazy Initialization
#             if self.database is None:
#                 print("Creating RealDatabase object...")
#                 self.database = ProxyPatternDemo.RealDatabase()
#
#             self.database.fetch_data()
#
#
# def main():
#     user = ProxyPatternDemo.DatabaseProxy("USER")
#     user.fetch_data()
#
#     print("--------------------------")
#
#     admin = ProxyPatternDemo.DatabaseProxy("ADMIN")
#     admin.fetch_data()
#
#     print("--------------------------")
#
#     # Same proxy reused (Real object already created)
#     admin.fetch_data()

File: C++/CreationalDesignPattern/ProxyPatternDemo.cpp
// struct Database {
//     virtual void fetchData() = 0;
//     virtual ~Database() = default;
// };
//
// // Real Subject
// struct RealDatabase : Database {
//     void fetchData() override {
//         std::cout << "Fetching confidential employee salary data..." << std::endl;
//     }
// };
//
// // Proxy
// class DatabaseProxy : public Database {
//     std::string role;
//     std::unique_ptr<RealDatabase> database;
//
// public:
//     explicit DatabaseProxy(std::string role) : role(std::move(role)) {}
//
//     void fetchData() override {
//         if (role != "ADMIN") {
//             std::cout << "Access Denied! Only ADMIN can access the database." << std::endl;
//             return;
//         }
//
//         // Lazy Initialization
//         if (!database) {
//             std::cout << "Creating RealDatabase object..." << std::endl;
//             database = std::make_unique<RealDatabase>();
//         }
//
//         database->fetchData();
//     }
// };
//
// int main() {
//     std::unique_ptr<Database> user = std::make_unique<DatabaseProxy>("USER");
//     user->fetchData();
//
//     std::cout << "--------------------------" << std::endl;
//
//     std::unique_ptr<Database> admin = std::make_unique<DatabaseProxy>("ADMIN");
//     admin->fetchData();
//
//     std::cout << "--------------------------" << std::endl;
//
//     // Same proxy reused (Real object already created)
//     admin->fetchData();
// }
